compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -446,10 +448,77 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                 }
             }
         }
+
         err.span_label(decl_span, "binding declared here but left uninitialized");
+        if show_assign_sugg {
+            struct LetVisitor {
+                decl_span: Span,
+                ty_span: Option<Span>,
+            }
+
+            impl<'v> Visitor<'v> for LetVisitor {
+                fn visit_stmt(&mut self, ex: &'v hir::Stmt<'v>) {
+                    if self.ty_span.is_some() {
+                        return;
+                    }
+                    if let hir::StmtKind::Local(hir::Local {
+                            span, init: None, ty: Some(ty), ..
+                        }) = &ex.kind && span.contains(self.decl_span) {
+                            self.ty_span = Some(ty.span);
+                    }
+                    hir::intravisit::walk_stmt(self, ex);
+                }
+            }
+
+            let mut visitor = LetVisitor { decl_span, ty_span: None };
+            visitor.visit_body(&body);
+            if let Some(ty_span) = visitor.ty_span {
+                self.suggest_assign_value(&mut err, moved_place, ty_span);
+            }
+        }
         err
     }
 
+    fn suggest_assign_value(
+        &self,
+        err: &mut Diagnostic,
+        moved_place: PlaceRef<'tcx>,
+        ty_span: Span,
+    ) {
+        let ty = moved_place.ty(self.body, self.infcx.tcx).ty;
+        debug!("ty: {:?}, kind: {:?}", ty, ty.kind());
+
+        let assign_value = match ty.kind() {
+            ty::Int(_) | ty::Uint(_) => format!("0"),
+            ty::Float(_) => format!("0.0"),
+            ty::Bool => format!("false"),
+            ty::Never | ty::Error(_) => "".to_string(),
+            ty::Adt(def, _substs) => {
+                if Some(def.did()) == self.infcx.tcx.get_diagnostic_item(sym::Vec) {
+                    format!("vec![]")
+                } else if let Some(default_trait) = self.infcx.tcx.get_diagnostic_item(sym::Default) &&
+                    self.infcx.tcx.infer_ctxt().enter(|infcx| {
+                        infcx.type_implements_trait(default_trait, ty, ty::List::empty(), self.param_env).may_apply()
+                    }) {
+                    format!("Default::default()")
+                } else {
+                    format!("todo!()")
+                }
```

Comment:
Formatting here is quite wonky, especially around the `type_implements_trait` check. I wonder if it would read better if some things were extracted into variables and such.

Maybe something along the lines of

```rust
let tcx = self.infcx.tcx;
let implements_default = |ty, param_env| {
    let Some(default_trait) = tcx.get_diagnostic_item(sym::Default) else {
        return false;
    };
    tcx.infer_ctxt().enter(|infcx| {
        infcx.type_implements_trait(default_trait, ty, ty::List::empty(), param_env).may_apply()
    })
};

match ty.kind() { 
    // ...
    ty::Adt(def, _) if Some(def.did()) == tcx.get_diagnostic_item(sym::Vec) => "vec![]",
    ty::Adt(_, _) if implements_default(ty, self.param_env) => "Default::default()",
    ty::Adt(_, _) => "todo!()",
}
```

Reply:
Yes, this is better in readability, thanks very much.